InheritedModel - an InheritedWidget for data models

HansMuller opened the pull request:
An `InheritedModel` is an InheritedWidget that's intended to be used as the base class for data models which are scoped to a widget tree.

A sample app that demonstrates defining a simple InheritedModel: https://gist.github.com/HansMuller/b947e76c34d941397989eaab537d45ca

HansMuller committed:
[WIP] Scoped inherited widget

Hixie commented:
The `dynamic`s are sad making.

Is there a good way to make this a subclass of InheritedWidget? It'd be nice if the regular InheritedWidgets didn't have to deal with any of this. A subclass would also give you somewhere to hang a `<T>` so you could remove some of the `dynamic`s (though admittedly not all), and would let you verify that the scope was null for regular widgets and not null for these.

Scope is a weird term for this (as you said in e-mail). Scope in this context is what I would use to describe the subtree of any InheritedWidget. Maybe "topic"?

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -3227,14 +3233,22 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
       _dependencies.add(ancestor);
-      ancestor._dependents.add(this);
+      // If we're ever called with scope == null (the common case) then
+      // InheritedElement.notifyClients will unconditionally rebuild all dependents.
```
This comment seems like it doesn't accurately describe what's going on here, in a way that actually quite dramatically changes my opinion of the patch. :-)

I think what this code is actually doing is better described by:
```
// If the `scope` is null, then always update this build context when
// the inherited widget changes, regardless of which scopes changed.
```

But then you also need to distinguish "we were once called with null" from "we have never been called" and from "we were once called with null then called with a scope" and "we were once called with a scope then called with null" (the last two should be the same).

(Side note: the documentation needs updating too.)

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1569,6 +1572,9 @@ abstract class InheritedWidget extends ProxyWidget {
   bool updateShouldNotify(covariant InheritedWidget oldWidget);
+
+  @protected
+  Set<dynamic> getUpdateScope(covariant InheritedWidget oldWidget) => _kEmptySet;
```
getUpdatedTopics maybe?

Hixie replied:
or updateShouldNotifyTopics?

Hixie commented:
How about `InheritedWidget` inherits from `InheritedWidgetBase<T>` with `T` set to `Null`?

HansMuller committed:
updated

krisgiesing commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1571,6 +1574,91 @@ abstract class InheritedWidget extends ProxyWidget {
 }
 
+abstract class InheritedModel<T> extends InheritedWidget {
+  const InheritedModel({ Key key, Widget child }) : super(key: key, child: child);
```
I think InheritedModel should store a reference to a T that represents the model, and take an argument that is the value of that reference.

That is, InheritedModel<Book> would have 

```
  const InheritedModel(this.model, { Key key, Widget child});
  Book model;
```

Hixie replied:
make sure to mention what `T` is in the docs when you write them (it's common for us to forget to document the generic types)

HansMuller replied:
T is the type of objects that represent dependencies.  I've covered T in the class doc I just added.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -45,6 +45,9 @@ export 'package:flutter/rendering.dart' show RenderObject, RenderBox, debugDumpR
 /// `children` property, and then provide the children to that widget.
 /// {@endtemplate}
 
+// TODO(hansmuller) - const UnmodifiableSetView<dynamic>.empty();
```
presumably we should deal with this todo before landing

Hixie commented:
I don't understand the distinction between "token" and "change".

krisgiesing reviewed (commented):
Ugh, I think I failed github 101 and many of my comments from Friday never got submitted.

krisgiesing commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1571,6 +1574,91 @@ abstract class InheritedWidget extends ProxyWidget {
+  bool updateShouldNotifyDependent(covariant InheritedModel<T> oldWidget, Set<T> changes);
+
+  @protected
+  T getChangeFor(dynamic token) => token;
```
The above comments were premised on T being the model type. It seems that's not the case, so perhaps I need to re-evaluate the types used throughout. I'll leave the comments intact for now, so you can see the direction of my thinking...

HansMuller replied:
T isn't the model type, it's the type of the objects used to track dependencies. 

The type of the model and how it's exposed is completely up to the InheritedModel subclass.

HansMuller committed:
checkpoint

HansMuller committed:
checkpoint

krisgiesing commented:
@apwilson is also likely interested in this change.

HansMuller committed:
checkpoint

HansMuller commented:
- Renamed `getChangeFor()` to `getDependencyFor()`. This value, which is of type T, represents the dependency between a descendant and the model.
- Fixed the case where depedencies depend on the model, i.e. where dependencies are not just a pass-through of the `inheritFromWidgetOfExactType` token parameter.
- Added an overall explanation to InheritedModel.

[This is a complete example that uses tokens as dependencies.](https://gist.github.com/HansMuller/a983d6e079be90947209dade80c46f3e)

[This is a complete example that uses a specialized dependency type.](https://gist.github.com/HansMuller/44163c7e4188dde32fefff8c711c55cb)

krisgiesing commented:
Thanks! I'll try to work up an example that uses the pattern we discussed offline, which is likely to look more like the second example than the first.

HansMuller commented:
@Hixie 
> I don't understand the distinction between "token" and "change".

"change" was a poor choice of names, it's "dependency" now.

A dependency is constructed by the InheritedModel from the token, so a dependency can depend on the model. Both the token and the dependency represent the the aspect of the model that an `inheritFromWidgetOfExactType` caller depends on.

I included some discussion of this in the class dartdoc.

Hixie commented:
I'm even more confused with the term "dependency". Isn't a dependency the actual link between the inherited widget and the, ah, dependent? Possibly scoped to a particular topic or subject?

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
````diff
@@ -1571,6 +1574,231 @@ abstract class InheritedWidget extends ProxyWidget {
+/// ```
+///
+/// Calling `MyModel.of(context, 'foo')` means that `context` depends on the
+/// `foo` aspect of `MyModel`.
````
Isn't it more accurate to say that it depends on MyModel, but only wants to be notified about changes relating to `foo`?
I could see calling the argument "aspect" or "subject" or some such. I'm not sure what `token` means in this context.

HansMuller commented:
> I'm even more confused with the term "dependency". Isn't a dependency the actual link between the inherited widget and the, ah, dependent? Possibly scoped to a particular topic or subject?

Yes, that's exactly what it is. In the simple case, the one where the "token" _is_ the dependency, the dependency is just a value that indicates what part/aspect/etc of the model the dependent widget depends on.  What values are valid and what they mean is up to the InheritedModel subclass.

By overriding getDependencyFor(token), apps can create more complex dependencies. Like the one in the "[complete example that uses a specialized dependency type](https://gist.github.com/HansMuller/44163c7e4188dde32fefff8c711c55cb)" where the dependency is essentially a model accessor and the current value of that accessor.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
````diff
@@ -1571,6 +1574,231 @@ abstract class InheritedWidget extends ProxyWidget {
+/// are created similarly, with the addition of a parameter that's
+/// use as the dependency token. For example:
+///
+/// ```
````
\`\`\`dart

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
````diff
@@ -1571,6 +1574,231 @@ abstract class InheritedWidget extends ProxyWidget {
+/// and decide if the model's changes correspond to any of the dependencies.
+///
+/// For example:
+/// ```
````
\`\`\`dart

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1571,6 +1574,231 @@ abstract class InheritedWidget extends ProxyWidget {
+///   @override
+///   bool updateShouldNotifyDependent(IntegerModel old, Set<PathDependency> dependencies ) {
+///     for (PathDependency dependency in dependencies) {
+///       if (dependency.value != integerTree.valueOf(dependency.path))
```
why would you want to do this instead of just looking up the path in the old model and comparing it to the value in the new model?

HansMuller replied:
Because the widget's model hasn't changed, only the model's internal state has changed.

Hixie replied:
I don't think we should be encouraging people to pass in mutable objects to InheritedWidgets. It's very easy to end up with hard-to-find bugs when doing that, because there's nothing that forces the inherited widget to get updated, making the whole update mechanism very brittle.

krisgiesing replied:
Can you elaborate? I thought InheritedWidget's reason for existing was to allow UI to depend on mutable state.

HansMuller committed:
checkpoint

HansMuller committed:
complete, with tests

HansMuller committed:
update

HansMuller committed:
workaround linter prefer_const_constructors bug

HansMuller commented:
Worked around Dart prefer_const_constructors linter bug, see  https://github.com/dart-lang/sdk/issues/34100

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1571,6 +1572,323 @@ abstract class InheritedWidget extends ProxyWidget {
+  bool updateShouldNotifyDependent(covariant InheritedModel<T> oldWidget, Set<T> dependencies);
+
+  // Return the first ancestor of context whose widget is type T and for which
+  // vistor(widget) returns false. If visitor is null then this function is the
```
typo in "vistor"

HansMuller committed:
updated

HansMuller commented:
Moved InheritedModel et al to a seprate file and replaced the private connections it depended on with public ones.

Added an `InheritedDependencies` class to framework.dart and three protected methods to InheritedElement:

```
  InheritedDependencies getDependencies(Element dependent) => _dependents[dependent];
  InheritedDependencies updateDependencies(Element dependent, Object aspect)
  void notifyDependent(covariant InheritedWidget oldWidget, Element dependent)
```

Removed the getDependencyFor() stuff for now. It could be added in a InheritedModel subclass. It isn't needed for the theme etc use cases.

The `inheritFromWidgetOfExactType` aspect argument is now Object instead of dynamic.

HansMuller committed:
update

krisgiesing commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4025,6 +4034,30 @@ class ParentDataElement<T extends RenderObjectWidget> extends ProxyElement {
 }
 
+/// Used by [InheritedElement] to represent the aspects of its
+/// [InheritedWidget] that will a dependent to be rebuilt.
```
will a dependent => will cause a dependent

HansMuller replied:
Thanks, fixed.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,212 @@
+  }) {
+    final InheritedElement model = _findModel<T>(context, visitor);
+    if (model != null)
+      return context.inheritFromWidgetOfExactType(T, aspect: aspect, target: model);
```
one problem with this approach is that if one of the widgets in the middle is updated to now have the aspect you're looking for, you're not going to get updated.

i.e. consider a tree A->B->C->D->E where A and C and inherited models M, only A provides aspect X, and E depends on model M aspect X. Now update the tree so that only B and C update, and C now has model X; E won't be notified and won't rebuild.

HansMuller replied:
That's true. Dynamically changing the list of aspects a model supports would require rebuilding its subtree. It's comparable to changing the tree by adding or removing an inherited widget.

Hixie replied:
I think we need to handle this case. It'll be really confusing if someone updates the theme but nothing happens until they happen to rebuild (e.g. by rotating the phone).

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -3227,14 +3236,14 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   @override
-  InheritedWidget inheritFromWidgetOfExactType(Type targetType) {
+  InheritedWidget inheritFromWidgetOfExactType(Type targetType, { Object aspect, InheritedElement target }) {
```
We should definitely assert that `target` is an ancestor.

HansMuller replied:
Good point, done.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,212 @@
+  bool contains(T dependency) => values.contains(dependency);
+
+  void add(T dependency) {
+    values.add(dependency);
```
assert not global?

HansMuller replied:
Yes, good idea.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4088,34 @@ class InheritedElement extends ProxyElement {
+  /// By default this method returns a constant [InheritedDependencies] object
+  /// for which [InheritedDependencies.isGlobal] is true.
+  @protected
+  InheritedDependencies updateDependencies(Element dependent, Object aspect) {
```
I'm a bit confused by this method being called `update` but returning a value. It took me a while to understand the model here. I'm not sure what to suggest, though...

Maybe if it took the current dependency as an argument?

HansMuller replied:
I could have explained these methods better, sorry about that. 

If there isn't a dependency (yet) then the "current dependency" is null: `getDependencies(dependent)` returns null. This method is responsible for constructing a a new dependency or updating an existing one. 

I guess this method's signature is more like the Map.update method than Element.update.

Hixie replied:
Map.update actually changes the map, though, which this method doesn't.

Maybe it would be less confusing if the InheritedDependencies object was immutable... though that would be a lot more expensive since you'd be copying a Set all the time...

> If there isn't a dependency (yet) then the "current dependency" is null

I think this is part of what is confusing. When it's null, this method is pure-functional, it just returns a new value, and the return value is critical. When it's not null, it's destructive, it actually changes the underlying value, and if the return value was ignored it would still work.

HansMuller replied:
I added a setDependencies() method which can be used by updateDependencies() :

```
  InheritedDependencies getDependencies(Element dependent) {
    return _dependents[dependent];
  }

  void setDependencies(Element dependent, InheritedDependencies value) {
    _dependents[dependent] = value;
  }

  void updateDependencies(Element dependent, Object aspect) {
    setDependencies(dependent, const InheritedDependencies()); // isGlobal => true
  }
```

Hixie commented:
Have you tested the performance impact?

HansMuller commented:
> Have you tested the performance impact?
I haven't checked to see if the Element changes have an impact  on performance. Will do that.

HansMuller committed:
update

HansMuller committed:
updated per review feedback

HansMuller committed:
update

HansMuller commented:
I checked the performance of flutter_gallery__transition_perf and complex_layout_scroll_perf__memory on a Moto E running Android 4.4.2.

There was no appreciable difference between this branch and the current tip of tree.

HansMuller committed:
Updated per review feedback

HansMuller commented:
To handle scenarios where a model ancestor is bypassed because it doesn't support the aspect its dependent cares about, I've simplified the API a little.

InheritedModel's now specify the aspects they support explicitly, with Set<T> valued property called "aspects". If it's null (the default), then the model supports all aspects.

The static inheritFrom<T ..>() method no longer uses a visitor callback parameter to decide which model ancestor to create a dependency on. It's now the just the first model of type T that supports the aspect the dependent specifies. The inheritFrom() logic actually creates additional dependencies on the model ancestors of type T below the one that supports the specified aspect. If one of these intermediate models is rebuilt with a changed set of aspects, then its subtree is rebuilt.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,247 @@
+    // Create a dependency on all of the type T ancestor models up until
+    // a model is found whose aspects are null or contain the given aspect.
+    // Rebuilding one of the intermediate models only causes its dependents
+    // to be rebuilt if its aspsects property changes (see _shouldNotify).
```
nit: typo on aspects

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1904,7 +1904,16 @@ abstract class BuildContext {
+  /// ancestor of this BuildContext's widget. The target is used instead of
+  /// the first [targetType] widget ancestor. This parameter is used by
+  /// [InheritedModel.inheritFrom]. It's typically not specified directly.
+  InheritedWidget inheritFromWidgetOfExactType(Type targetType, { Object aspect, InheritedElement target });
```
can we refactor this so instead of one method that does the walk if target is null and then depends on target, we have two methods, one for doing the walk and one for actually setting up the inheritance? i think that would be cleaner. It's weird to say "inheritFromWidgetOfExactType" and then provide the target (at which point the type isn't even relevant?).

HansMuller replied:
I agree that it would cleaner to just a separate method for this case.

The reason I'd done things this way was to ensure that the checking done by StatefulElement.inheritFromWidgetOfExactType() would happen, even if the target was specified. The big error message there would be a little harder to understand if it was generated by a new method rather than the one the user called directly.

I'll take a crack at refactoring things and dealing with StatefulElement's error reporting.

HansMuller commented:
> One of the things that would be nice to do is to have it so that subclasses don't have to subclass the element as well. We could do this by putting updateDependencies on the widget rather than the element, and passing it an interface that has the get and set dependency methods on it (implemented by the InheritedElement, much like how we do this with BuildContext vs Element).

I believe that you're referring to InheritedWidget subclasses like InheritedModel, not InheritedModel subclasses.

To get rid of the need for an InheritedElement subclass we'd also to add notifyDependent() to InheritedWidget. I think this particular set of factorings might complicate matters as much as cleaning them up. And some InheritedWidget subclasses might still need to subclass InheritedElement anyway, e.g. to do some work after notfiyClients() has finished.

> Another thing we could do is remove the InheritedDependencies class from framework.dart, since framework.dart doesn't actually care about it (e.g. it doesn't use isUniversal). For example, if we introduced a base class for InheritedWidget/InheritedModel that was generic over the type of object that gets put into the dependency map, we could make InheritedWidget override it with T=Null, and have it not bother with any objects in that map at all, just storing nulls.

This sounds reasonable. I will see if I can thread the needle.

HansMuller committed:
removed the InheritedDependencies class

HansMuller committed:
removed inheritWidgetFromExactType target parameter

HansMuller committed:
update

HansMuller committed:
fixed a typo

HansMuller commented:
Removed the InheritedDependencies type. Dependencies are now just objects (null by default). InheritedModel uses sets.

Removed the inheritFormWidgeOfExactType target parameter and added BuildContext.inheritFromElement().

Moved the StateElement error checking inheritFormWidgeOfExactType to inheritFromElement and updated the error report just a little.

krisgiesing commented:
The latest series of changes appears to have dropped an earlier feature that my prototype relies on. Specifically I would like to retrieve an ancestor using inheritFromWidgetOfExactType, then use the result to determine what aspect of the model to depend on (because the returned ancestor has a reference to the model, which I can use to look up the aspect).  Previously this decoupling was achieved through getDependencyFor() but that method has been removed. Is there another way to achieve the same effect? I haven't found one.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1705,7 +1705,7 @@ class _InactiveElements {
       elements.reversed.forEach(_unmount);
     } finally {
-      assert(_elements.isEmpty);
+       assert(_elements.isEmpty);
```
nit: extraneous space

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1870,6 +1870,17 @@ abstract class BuildContext {
   /// render object is usually short.
   Size get size;
 
+  /// Registers this build context with [ancestor] such that when that when
```
nit: duplicate "that when"

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1870,6 +1870,17 @@ abstract class BuildContext {
+  /// the appropriate [InheritedElement] ancestor.
+  ///
+  /// All of the qualifications about when [inheritFromWidgetOfExactType] can
+  /// be called, apply to this method as well.
```
nit: i think that comma is extraneous?

HansMuller replied:
Yes, it is.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1879,9 +1890,9 @@ abstract class BuildContext {
+  /// This method should not be called from widget constructors or from
   /// [State.initState] methods, because those methods would not get called
-  /// again if the inherited value were to change. To ensure that the widget
+  /// again if the inherited value was to change. To ensure that the widget
```
why "was"?

HansMuller replied:
Not sure what I were thinking here. Reverted.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -3227,15 +3243,31 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
+    assert(ancestor != null);
+    assert(() {
+      if (_parent == null) {
+        // We're being deactivated, see deactivateChild()
```
can you `assert(_debugLifecycleState == _ElementLifecycle.inactive)` in this case?

HansMuller replied:
It's not necessarily inactive (yet) at this point.

The element's _parent is null if InheritFromWidgetOfExactType is called from a State.deactivate() override. FormFieldState does this:

```
  @override
  void deactivate() {
    Form.of(context)?._unregister(this);
    super.deactivate();
  }
```

The `_unregister()` method is trying to remove the field from its form's list of fields.

The deactivate() method is called by deactivateChild() after the child's parent is set to null but before _debugLifecycleState is changed to  _ElementLifecycle.inactive at the end of Element.deactivate():

```
  void deactivate() {
    assert(_debugLifecycleState == _ElementLifecycle.active);
    ...
    assert(() { _debugLifecycleState = _ElementLifecycle.inactive; return true; }());
  }
```

Hixie replied:
I wonder if we could add another lifecycle state for this intermediate mode...

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1870,6 +1870,17 @@ abstract class BuildContext {
+  ///
+  /// All of the qualifications about when [inheritFromWidgetOfExactType] can
+  /// be called, apply to this method as well.
+  InheritedWidget inheritFromElement(InheritedElement ancestor, { Object aspect });
```
these docs should mention and justify the return value

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4089,87 @@ class InheritedElement extends ProxyElement {
+  /// [getDependencies].
+  ///
+  /// By default this method sets the inherited dependencies for [dependent]
+  /// to null. Subclasses can manage their own dependencies values so that they
```
after "null", explain the implications of this

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4089,87 @@ class InheritedElement extends ProxyElement {
+  ///
+  /// Each dependent element is mapped to a single object value
+  /// which represents how the element depends on this
+  /// [InheritedElement]. This value is null by default and by default
```
maybe `and` -> `so that`?

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4089,87 @@ class InheritedElement extends ProxyElement {
+  /// Subclasses can manage these values with [updateDependencies]
+  /// so that they can selectively rebuild dependents in [notifyDependents].
+  ///
+  /// See also:
```
should probably say something to the effect that this should typically only be called in overrides of updateDependencies.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4089,87 @@ class InheritedElement extends ProxyElement {
+  /// so that they can selectively rebuild dependents in
+  /// [notifyDependents].
+  ///
+  /// See also:
```
should probably say something to the effect that this should typically only be called in overrides of updateDependencies.

Hixie commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -4055,6 +4089,87 @@ class InheritedElement extends ProxyElement {
+  ///  * [notifyDependent], which can be overridden to use a dependent's
+  ///    dependencies value to decide if the dependent needs to be rebuilt.
+  @protected
+  void updateDependencies(Element dependent, Object aspect) {
```
maybe point to InheritedModel as an example of a class that uses this

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+import 'framework.dart';
+
+/// An [InheritedWidget] that's intended to be used as the base class for data
+/// models which are scoped to a widget tree.
```
"data models" isn't defined at this point, maybe replace the term here with something more explicit?

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+/// models which are scoped to a widget tree.
+///
+/// An inherited widget's dependents are unconditionally rebuilt when the
+/// inherited widget changes per [InheritedWidget.updateShouldNotify].
```
it's not clear here that this is not what this widget does. Maybe `An inherited widget's dependents are normally unconditionally rebuilt when the inherited widget changes per [InheritedWidget.updateShouldNotify]. This widget changes this to only call...` or some such

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+/// class MyModel extends InheritedModel<String> {
+///   // ...
+///   static MyModel of(BuildContext context, String aspect) {
+///     return InheritedModel.inheritFrom<MyModel>(context, aspect: aspect);
```
wouldn't this only return the "aspect" subpart of MyModel, typically?

HansMuller replied:
Yes, that would probably work well. Although in the interest of keeping this example as simple as possible it might be worth leaving that out.

Once we've accumulated some experience using this class beyond tests, this example should merit a revision.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+///
+/// Calling `MyModel.of(context, 'foo')` means that `context` should only
+/// be rebuilt when the `foo` aspect of `MyModel` changes. If the aspect
+/// is null, then the the model supports all aspects.
```
nit: duplicate "the"

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+      if (model == lastModel)
+        return value;
+    }
+    assert(false, 'Unreachable');
```
we usually just assert(false) without a comment.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+      assert(aspect is T);
+      setDependencies(dependent, dependencies == null
+        ? (new HashSet<T>()..add(aspect))
+        : dependencies..add(aspect));
```
this might be cleaner as:

`(dependencies ?? new HashSet<T>())..add(aspect)`

HansMuller replied:
Yes, that's tidier

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+  }
+}
+
+bool _notEqual<T>(Set<T> a, Set<T> b) {
```
You can use setEquals from the foundation library

HansMuller replied:
Thanks, I didn't know of setEquals.

Hixie commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -6,6 +6,7 @@ import 'package:flutter_test/flutter_test.dart';
 import 'package:flutter/material.dart';
 
 void main() {
+  /*
```
i suspect this was accidental

Hixie commented:
This seems reasonable. Mostly nits. I'm dubious about the `aspects` argument, I suspect we can do equally well without it.

HansMuller committed:
Updated per review feedback

HansMuller commented on packages/flutter/lib/src/widgets/inherited_model.dart:
```diff
@@ -0,0 +1,234 @@
+    });
+    if (modelParent == null)
+      return;
+    _findModels<T>(modelParent, aspect, result);
```
That is a little nicer.

HansMuller committed:
updated per review feedback

HansMuller committed:
updated _findModels()

HansMuller committed:
removed InheritedModel.aspects

HansMuller commented:
Removed the explicit support for models that only support a specific set of aspects. Subclasses can provide this feature, which is useful for models that can partially shadow their ancestors, by overriding the `isSupportedAspect` method. The inherited model test does this.

Hixie commented on packages/flutter/lib/src/widgets/inherited_model.dart:
````diff
@@ -0,0 +1,197 @@
+///   // ...
+/// }
+/// ```
+///
````
currently the code above is not analyzed, because it's not in a "## Sample code" section. Dunno if you are concerned about that or not. (The advantage of having it in such a section is it'll get updated when we do things like introduce new lints, e.g. the upcoming "remove new" lint.)

HansMuller committed:
updated

Hixie commented:
![LGTM](http://junkyard.damowmow.com/578)
